### DISCUSSION TITLE
[RFC] Automatic Font Fallback for layoutshiftless font-display: swap

### DISCUSSION THREAD

**Discussion by patrickhulce** [vercel/next.js#24438] [Ideas]
### Describe the feature you'd like to request

I would like next.js to provide fallback fonts by default that enable use of `font-display: swap` for performance benefits without incurring the cost of layout shifts.

**Problem**
- You want to use a webfont on users' first page experience.
- You don't want waiting on the font to block FCP.
- You don't want layout shifts.

Today you have to pick two of the three, OR you have to painstakingly use a tool like [the font style matcher](https://meowni.ca/font-style-matcher/) to find matching font settings, apply a class that sets the overrides for the fallback fonts, load the fonts with the FontFace API to observe readiness, and then trigger the upgrade when the webfont has loaded.

I want an end result that works like [this example](https://codepen.io/patrickhulce/pen/LYxgOpo) out-of-the-box for users of common webfonts.

### Describe the solution you'd like

[Next.js font optimizations](https://github.com/vercel/next.js/pull/14746) already introduced the concept of modifying and optimizing font declarations from trusted parties. I propose extending this concept to building a database of known matching fallback fonts that could be applied using the new overriding font descriptors as part of font optimization. If taken on before the `advance-override` descriptors are available, it would entail injecting the utility classes and observing font availability to apply body-wide classes as well (Much more work).

The usage pattern for next.js users should be largely invisible and happen by default with a flag to disable (or enable during experimental phase).

### Describe alternatives you've considered

- Ignore `advance-override` and other font descriptors and stick to `letter-spacing` workarounds to be implemented at a CSS framework level like tailwind (still an option, though probably more limited in scope and more difficult to use, limits total possible impact)
- Go straight to the source and have Google Fonts/TypeKit/etc provide sensible fallbacks for every font they offer. (this is ideal, but I'm assuming they'd want validation of the approach before doing so, this seems like a great place to do that 😃 )
- Do nothing, live with the CLS, FCP impact, and/or brand impact of the other font-display options.

**Related Issues**
Customizable font display with inlining optimization (https://github.com/vercel/next.js/issues/21555)
Dynamic font loading (https://github.com/vercel/next.js/issues/20134)
Async font styles (https://github.com/vercel/next.js/issues/16065)

**Related Resources**
[Reducing layout shift from fonts article](https://simonhearne.com/2021/layout-shifts-webfonts/#reduce-layout-shift-with-f-mods)
[font-display: swap w/o layout shift example](https://codepen.io/patrickhulce/pen/LYxgOpo)
[Chromium issue for `advance-override`](https://bugs.chromium.org/p/chromium/issues/detail?id=1137633&q=font&can=5)
[CSS WG discussion on override inline spacing](https://github.com/w3c/csswg-drafts/issues/5533)
[CSS WG discussion on overrides for fallback fonts](https://github.com/w3c/csswg-drafts/issues/126)

**Comment by rauchg** [comment]
The quintessential "have your webfont cake and eat it too". I love this. Thanks @patrickhulce for sharing your proposal!

**Comment by liamquin** [comment]
There are probably tens of thousands of typefaces in at least moderately widespread use on the Web. I know someone who has over 80,000 fonts on their system for testing purposes.

However, if browsers had both a fallback font and also stored for each typeface deviations from that fallback font's metrics, it might be possible to simulate the font well enough to avoid reflow in the vast majority of cases without resorting to uneven letter-spacing, which can be an accessibility, comprehension and retention barrier, and can be very off-putting. It's also necessary to consider the case that the font never actually downloads - e.g. a firewall blocks Web fonts.

There were attempts to do something similar both for PDF and for word processing documents in the late 1980s and 1990s. “Panose” numbers were an attempt, too, to automate fallback. However, i am not certain that Panose numbers really covered the needs of fallback beyond Latin scripts very well; any new mechanism would need to work everywhere, for all languages, of course.

It might actually be worth considering what would happen if browsers shipped with a pre-loaded cache of the most common few thousand Web fonts. Or at least font metrics? Metrics can change between font versions, but it's rare.
